Allow WKNavigationAction.request.URL to be valid NSURL that isn't valid WTF::URL

achristensen07 opened the pull request:
#### 7b37bf4eada1c881ba0453f9dc6b04b772a03e47
<pre>
Allow WKNavigationAction.request.URL to be valid NSURL that isn&apos;t valid WTF::URL
<a href="https://bugs.webkit.org/show_bug.cgi?id=294333">https://bugs.webkit.org/show_bug.cgi?id=294333</a>
<a href="https://rdar.apple.com/145790356">rdar://145790356</a>

Reviewed by Chris Dumez.

In 289946@main I made it so an invalid WTF::URL can&apos;t convert to a valid NSURL.
This fixed a large number of issues involving improper use of URLs.
However, this also prevented a legitimate use of invalid WTF::URLs:
An app can get a WKNavigationAction with the URL and decide to do something with
it that is not related to WebKit.  For compatibility, we need to explicitly allow
this.  Since the WebCore::ResourceRequest IPC encoder uses conversion to NSURL,
if we have an invalid WTF::URL we need to send the invalid URL string separately
then use it on the other side.  Since NSURL.URLWithString also doesn&apos;t successfully
parse such a URL, we need to use CFURLCreateAbsoluteURLWithBytes, which I made
into its own function to share between its existing and introduced uses.

An alternative I considered was to percent encode spaces in hosts of URLs with
non-special schemes, but that goes against behavior of Chrome, Firefox, Safari,
and NSURL.URLWithString so I decided not to.

* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL):
(WTF::URL::createCFURL const):
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction request]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJS):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadInvalidURLRequest.mm:
(TestWebKitAPI::TEST(WebKit, NavigateToInvalidURL)):

Canonical link: <a href="https://commits.webkit.org/296130@main">https://commits.webkit.org/296130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/052587e1af40cc973ae180e965be5b136fb2f2ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107473 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112687 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58009 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109437 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35655 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81581 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96860 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61966 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21485 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15001 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57453 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100062 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91415 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115788 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106020 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25450 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90619 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34915 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93116 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23039 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35277 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13053 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30285 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34461 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40003 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130336 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34207 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35446 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37562 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9e4881ce8165a77af58e0eee5c1ffdc2d18ac05c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46611

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d79de5f6c4cf24454f38e2221f159880ce6549df. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46611

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/17860bf396e1335936ff8a18dbd74c51a74bce21. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46611

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/759feeddad5593493933f67073a7390e0c39f9ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46611

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/052587e1af40cc973ae180e965be5b136fb2f2ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46611

achristensen07 committed:
Allow WKNavigationAction.request.URL to be valid NSURL that isn't valid WTF::URL
https://bugs.webkit.org/show_bug.cgi?id=294333
rdar://145790356

Reviewed by Chris Dumez.

In 289946@main I made it so an invalid WTF::URL can't convert to a valid NSURL.
This fixed a large number of issues involving improper use of URLs.
However, this also prevented a legitimate use of invalid WTF::URLs:
An app can get a WKNavigationAction with the URL and decide to do something with
it that is not related to WebKit.  For compatibility, we need to explicitly allow
this.  Since the WebCore::ResourceRequest IPC encoder uses conversion to NSURL,
if we have an invalid WTF::URL we need to send the invalid URL string separately
then use it on the other side.  Since NSURL.URLWithString also doesn't successfully
parse such a URL, we need to use CFURLCreateAbsoluteURLWithBytes, which I made
into its own function to share between its existing and introduced uses.

An alternative I considered was to percent encode spaces in hosts of URLs with
non-special schemes, but that goes against behavior of Chrome, Firefox, Safari,
and NSURL.URLWithString so I decided not to.

* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/cf/URLCF.cpp:
(WTF::URL::createCFURL):
(WTF::URL::createCFURL const):
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationAction.mm:
(-[WKNavigationAction request]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJS):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadInvalidURLRequest.mm:
(TestWebKitAPI::TEST(WebKit, NavigateToInvalidURL)):

Canonical link: https://commits.webkit.org/296130@main
